### PR TITLE
Alt + Enter sends current line to FSI if there is no selection

### LINKED ISSUE #256: Alt + Enter should send current line to F# Interactive if there is no selection
It's a relatively small change with good effects.

Some benefits:

 - The behaviour is consistent with that of FSharpBinding in Xamarin Studio / MonoDevelop.
 - A number of people have requested this e.g. http://vfpt.uservoice.com/forums/247560-general/suggestions/5737083-make-altenter-send-the-current-line-to-fsi-when-no

I'll send a PR for discussion shortly.

### PR DESCRIPTION
**Pull request by dungpa**
Close #256.

### CODE REVIEW

File: vsintegration/src/vs/FsPkgs/FSharp.VS.FSI/fsiSessionToolWindow.fs [hunk 1]
```diff
@@ -455,16 +455,17 @@ type internal FsiToolWindow() as this =
             let dte = provider.GetService(typeof<DTE>) :?> DTE        
             let activeD = dte.ActiveDocument            
             match dte.ActiveDocument.Selection with
-            | :? TextSelection as selection ->
+            | :? TextSelection as selection when selectLine || selection.Text = "" ->
                 let origLine = selection.CurrentLine
```
Comment (latkin) [inline vsintegration/src/vs/FsPkgs/FSharp.VS.FSI/fsiSessionToolWindow.fs]: Is this guy used? Looks like it can be removed.

**Comment by latkin** [comment]
LGTM

File: vsintegration/src/vs/FsPkgs/FSharp.VS.FSI/fsiSessionToolWindow.fs (hunk 1 above)
Comment (dungpa) [inline vsintegration/src/vs/FsPkgs/FSharp.VS.FSI/fsiSessionToolWindow.fs]: Thanks. It has been removed.

**Comment by forki** [comment]
Awesome. One of the fastest pull request, ever. Kudos

**Comment by latkin** [comment]
spaceballs-ludicrous-speed-o

**Comment by rojepp** [comment]
Great work! Should the 'Send line to F# interactive' be removed?

**Comment by forki** [comment]
Yes I think it should.

**Comment by dungpa** [comment]
@rojepp @forki Done via #260.

**Comment by toburger** [comment]
Is this executing a single line or the current expression? Wouldn't it make even more sense to execute the current expression (like in LISP)?

The reason why I find that more practical is the following:
I usually tend to iteratively write a function (which can get longer/more complex by the time) and I'm directly executing it for testing.
```fsharp
let doSomethingWith x =
   // some more lines...
   x + x

doSomethingWith 1
```

Currently I have to select the whole code block (from let doSomething... to the last line) and then ```alt+enter```.
When ```alt+enter``` evaluates the current expression I can go to the last line of the expression (or from where I'm currently editing the expression) and press ```alt+enter``` go down to my test code and press ```alt+enter``` again.

Maybe there is also another practical use with modules:
```fsharp
module A =
    type B = B of int
    let foo (B b) = ()
    let bar (B b) = ()
```
Sometimes it becomes very cumbersome to select the whole module when I only want to change binding x and with nested types the problem gets even worse because when nested types are used elsewhere the types don't match anymore (so I have to re-evaluate also all other expressions). So ```alt+enter``` should then only evaluate (the other expressions remain untouched)
```fsharp
module A =
   let bar (B b) = ()
```
I hope I could make my point.

**Comment by rojepp** [comment]
@toburger I think what you're asking for would be hard to accomplish. Specifically as a user, this would make me unsure of what I was about to execute in FSI. However, the need may be addressed by adding a command for expanding selection to next expression level. The idea that each invocation expands current selection to the expression it is in. You could then repeat that command to further expand the selection and when you're happy, you send it to FSI. Maybe a good idea for [VFPT](https://vfpt.uservoice.com/)?

**Comment by forki** [comment]
Yes "expanding selection" would also be useful for #244

**Comment by dungpa** [comment]
I agree with @rojepp. The selection should be made explicit before sending to FSI.

AFAIK ReSharper has extend/shrink selection feature. It would be even more useful for F#, especially when working with FSI.

Could anyone create an suggestion at https://vfpt.uservoice.com/? We could trial an implementation there first.

**Pull request closed without merging**